File: src/utils/prompt.ts
/**
 * Interactive prompt utilities
 */

import { runtime } from "../runtime/index.ts";

/**
 * Read a single line of input from the user
 */
async function readLine(): Promise<string> {
  const buf = new Uint8Array(1024);
  const n = await runtime.io.stdin.read(buf);
  const isInputReceived = n !== null;
  if (isInputReceived) {
    return new TextDecoder().decode(buf.subarray(0, n)).trim();
  }
  return "";
}

/**
 * Display a Y/n confirmation prompt
 * @param message Confirmation message
 * @returns true if user selects Yes, false otherwise
 */
export async function confirm(message: string): Promise<boolean> {
  // VIBE_FORCE_INTERACTIVE: When set to "1", forces the CLI to treat stdin as interactive.
  // This is necessary for E2E testing with node-pty, which creates a pseudo-terminal (PTY)
  // that Deno.stdin.isTerminal() doesn't recognize as a true TTY. Without this flag,
  // interactive prompts would fail in E2E tests even though they're running in a PTY.
  const forceInteractive = runtime.env.get("VIBE_FORCE_INTERACTIVE") === "1";
  const isInteractive = forceInteractive || (runtime.io.stdin.isTerminal());

  // In non-interactive environments (CI, scripts), automatically return false
  if (!isInteractive) {
    console.error(
      "Error: Cannot run in non-interactive mode with uncommitted changes.",
    );
    return false;
  }

  while (true) {
    // Use writeSync to ensure immediate output in PTY environments (bypasses buffering)
    runtime.io.stderr.writeSync(new TextEncoder().encode(`${message}\n`));
    const input = await readLine();

    const isYes = input === "Y" || input === "y" || input === "";
    if (isYes) {
      return true;
    }

    const isNo = input === "N" || input === "n";
    if (isNo) {
      return false;
    }

    runtime.io.stderr.writeSync(new TextEncoder().encode("Invalid input. Please enter Y/y/n/N.\n"));
  }
}

/**
 * Display a numbered selection prompt
 * @param message Prompt message
 * @param choices Array of choices
 * @returns Selected index (0-based)
 */
export async function select(
  message: string,
  choices: string[],
): Promise<number> {
  // VIBE_FORCE_INTERACTIVE: When set to "1", forces the CLI to treat stdin as interactive.
  // This is necessary for E2E testing with node-pty, which creates a pseudo-terminal (PTY)
  // that Deno.stdin.isTerminal() doesn't recognize as a true TTY. Without this flag,
  // interactive prompts would fail in E2E tests even though they're running in a PTY.
  const forceInteractive = runtime.env.get("VIBE_FORCE_INTERACTIVE") === "1";
  const isInteractive = forceInteractive || (runtime.io.stdin.isTerminal());

  // In non-interactive environments, throw an error (select() requires interaction)
  if (!isInteractive) {
    throw new Error("Cannot run select() in non-interactive mode");
  }

  while (true) {
    // Use writeSync to ensure immediate output in PTY environments (bypasses buffering)
    runtime.io.stderr.writeSync(new TextEncoder().encode(`${message}\n`));
    for (let i = 0; i < choices.length; i++) {
      runtime.io.stderr.writeSync(new TextEncoder().encode(`  ${i + 1}. ${choices[i]}\n`));
    }
    runtime.io.stderr.writeSync(new TextEncoder().encode("Please select (enter number):\n"));

    const input = await readLine();
    const number = parseInt(input, 10);

    const isValidNumber = !isNaN(number) && number >= 1 && number <= choices.length;
    if (isValidNumber) {
      return number - 1;
    }

    runtime.io.stderr.writeSync(
      new TextEncoder().encode(
        `Invalid input. Please enter a number between 1 and ${choices.length}.\n`,
      ),
    );
  }
}

File: rust/crates/vibe-core/src/prompt.rs
//! Interactive confirm / select prompts.
//!
//! Ported from `packages/core/src/utils/prompt.ts`. The trait lets `jump` (and
//! later commands) be tested with a scripted [`FakePrompt`]; [`RealPrompt`]
//! drives the real terminal through the injected [`Io`]. Interactivity is gated
//! the same way as the TS: forced on by `VIBE_FORCE_INTERACTIVE=1`, otherwise it
//! requires a stdin tty.

use crate::error::{Result, VibeError};
use crate::io::Io;

/// Y/n confirmation and numbered selection prompts.
pub trait Prompt {
    /// Show a Y/n prompt; `true` for yes. Non-interactive → `false`.
    fn confirm(&self, message: &str) -> bool;

    /// Show a numbered choice list; returns the selected 0-based index.
    /// Non-interactive → `Err` (matches the TS `throw`).
    fn select(&self, message: &str, choices: &[String]) -> Result<usize>;
}

/// Production [`Prompt`] over the injected [`Io`].
pub struct RealPrompt<'a, I: Io> {
    io: &'a I,
}

impl<'a, I: Io> RealPrompt<'a, I> {
    pub fn new(io: &'a I) -> Self {
        RealPrompt { io }
    }

    fn is_interactive(&self) -> bool {
        self.io.force_interactive() || self.io.is_stdin_terminal()
    }
}

impl<I: Io> Prompt for RealPrompt<'_, I> {
    fn confirm(&self, message: &str) -> bool {
        if !self.is_interactive() {
            // Non-interactive: auto-decline, matching prompt.ts.
            self.io.writeln_stderr(
                "Error: Cannot run in non-interactive mode with uncommitted changes.",
            );
            return false;
        }

        loop {
            self.io.writeln_stderr(message);
            let Some(input) = self.io.read_line() else {
                return false; // EOF → No.
            };

            // Empty or starts with y/Y → yes; n/N → no; else reprompt.
            let is_yes = input.is_empty() || input == "Y" || input == "y";
            if is_yes {
                return true;
            }
            let is_no = input == "N" || input == "n";
            if is_no {
                return false;
            }
            self.io
                .writeln_stderr("Invalid input. Please enter Y/y/n/N.");
        }
    }

    fn select(&self, message: &str, choices: &[String]) -> Result<usize> {
        if !self.is_interactive() {
            return Err(VibeError::Argument(
                "Cannot run select() in non-interactive mode".to_string(),
            ));
        }

        loop {
            self.io.writeln_stderr(message);
            for (i, choice) in choices.iter().enumerate() {
                self.io.writeln_stderr(&format!("  {}. {choice}", i + 1));
            }
            self.io.writeln_stderr("Please select (enter number):");

            let Some(input) = self.io.read_line() else {
                // EOF → last choice (usually Cancel), matching prompt.ts.
                return Ok(choices.len().saturating_sub(1));
            };

            let parsed = input.parse::<usize>().ok();
            let is_valid = parsed.is_some_and(|n| n >= 1 && n <= choices.len());
            if is_valid {
                return Ok(parsed.unwrap() - 1);
            }
            self.io.writeln_stderr(&format!(
                "Invalid input. Please enter a number between 1 and {}.",
                choices.len()
            ));
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::io::FakeIo;

    fn prompt_io(stdin: &[&str], interactive: bool) -> FakeIo {
        FakeIo::new().with_stdin(stdin).stdin_tty(interactive)
    }

    #[test]
    fn confirm_yes_on_empty_or_y() {
        for line in ["", "y", "Y"] {
            let io = prompt_io(&[line], true);
            assert!(RealPrompt::new(&io).confirm("ok?"));
        }
    }

    #[test]
    fn confirm_no_on_n() {
        let io = prompt_io(&["n"], true);
        assert!(!RealPrompt::new(&io).confirm("ok?"));
    }

    #[test]
    fn confirm_reprompts_on_invalid_then_accepts() {
        let io = prompt_io(&["maybe", "y"], true);
        assert!(RealPrompt::new(&io).confirm("ok?"));
        assert!(io.stderr_text().contains("Invalid input"));
    }

    #[test]
    fn confirm_eof_is_no() {
        let io = prompt_io(&[], true);
        assert!(!RealPrompt::new(&io).confirm("ok?"));
    }

    #[test]
    fn confirm_non_interactive_is_no() {
        let io = prompt_io(&["y"], false); // tty off, no force.
        assert!(!RealPrompt::new(&io).confirm("ok?"));
    }

    #[test]
    fn confirm_force_interactive_overrides_tty() {
        let io = FakeIo::new()
            .with_stdin(&["y"])
            .stdin_tty(false)
            .with_env("VIBE_FORCE_INTERACTIVE", "1");
        assert!(RealPrompt::new(&io).confirm("ok?"));
    }

    #[test]
    fn select_returns_zero_based_index() {
        let io = prompt_io(&["2"], true);
        let choices = vec!["a".into(), "b".into(), "Cancel".into()];
        assert_eq!(RealPrompt::new(&io).select("pick", &choices).unwrap(), 1);
    }

    #[test]
    fn select_reprompts_out_of_range() {
        let io = prompt_io(&["9", "1"], true);
        let choices = vec!["a".into(), "Cancel".into()];
        assert_eq!(RealPrompt::new(&io).select("pick", &choices).unwrap(), 0);
    }

    #[test]
    fn select_eof_returns_last_choice() {
        let io = prompt_io(&[], true);
        let choices = vec!["a".into(), "b".into(), "Cancel".into()];
        assert_eq!(RealPrompt::new(&io).select("pick", &choices).unwrap(), 2);
    }

    #[test]
    fn select_non_interactive_errors() {
        let io = prompt_io(&["1"], false);
        let choices = vec!["a".into(), "Cancel".into()];
        assert!(RealPrompt::new(&io).select("pick", &choices).is_err());
    }
}
